pipeline/authz/remote.go

```diff
@@ -67,7 +67,24 @@ func (a *AuthorizerRemote) GetID() string {
 func (a *AuthorizerRemote) Authorize(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, rl pipeline.Rule) (err error) {
 	ctx, span := a.tracer.Start(r.Context(), "pipeline.authz.AuthorizerRemote.Authorize")
 	defer otelx.End(span, &err)
-	r = r.WithContext(ctx)
+
+	// Issue #1136 - Authorizer "remote" throws exception "invalid Read on closed Body" if request body is present in request
+	// Original code
+	//r = r.WithContext(ctx)
+	//
+	// Problem with the original code is when r.WithContext is called, it creates a new request with the same body as the original request with a modified context.
+	// r is a pointer and when its assigned to the new request message it scope becomes local to this function.
+	// The calling function expects the original request message to be modified as the message is processed by the pipeline.
+	// When message processing is complete the calling function does not have access to the modified http message (r)
+	//
+	// There are several possible solutions:
+	// 1. Create a function in http.request to .SetContext(ctx) which replaces context but does not create new http.Request instance
+	// 2. Modify the Authorize function to return a new http.Request pointer used by this function during message processing
+	// 3. Assign the new request to the original request pointer
+	//
+	// The change below implements option 3
+	*r = *(r.WithContext(ctx))
```

Comment:
I'm probably wrong but this looks odd to me. `WithContext` creates a shallow clone of the request which might be the actual issue, given how `r` is used below. Have you tried `r = r.Clone(ctx)` instead?

Reply:
Hi taishp.. Yes I tried Clone() also. Using either function r.WithContext(ctx) or r.Clone(ctx) create a new request message which gets assigned to r (HTTP Request *). The calling function to Authorizer expects r* it passed to persist throughout the pipeline as it will use the context and payload for further processing. Once we assign r (via WithContext or Clone) in the Authorize function, it's pointing to a new HTTP Request * which goes out of scope when the Authorize function returns. 

It appears this issue was introduced in #1079 when adding tracing improvements. 
While testing, if we remove the lines added from #1079:
	ctx, span := a.tracer.Start(r.Context(), "authz.remote")
	defer otelx.End(span, &err)
	r = r.WithContext(ctx)
The issue does not occur as the scope of r does not change.

As another test, I added a function SetContext(ctx) to http.request to assign the context to the request messaeg but not create a new http.http request. This also worked as again the scope of r persisted.

Reply:
To address the syntax looking odd, here is another datapoint that follows the same pattern as this code change. Line 133 calls HandleRequest which calls the remote::Authorize function where the fix was made. In the below code snippet we confirm that the r.Out scope must be retained throughout the message processing.

proxy.go 
132	*r.Out = *r.Out.WithContext(context.WithValue(r.Out.Context(), ContextKeyMatchedRule, rl))
133	s, err := d.r.ProxyRequestHandler().HandleRequest(r.Out, rl)
134	if err != nil {
135		*r.Out = *r.Out.WithContext(context.WithValue(r.Out.Context(), director, err))
136		return
137	}
138	*r.Out = *r.Out.WithContext(context.WithValue(r.Out.Context(), ContextKeySession, s))

Reply:
Just enhanced and added to pull the unit testing in remote_test.go. To test this failure the message body sent to the authorize function is compared to the message body after the authorize pipeline completes. There are two tests that already pass a message body into the Authorize function and these were both verified to fail without 1136 fix and pass after 1136 fix.